:book: upgrade migration guide for v3 (#1916)
kubernetes-sigs/kubebuilder: docs/book/src/migration/manually_migration_guide_v2_v3.md

Migrating between project configuration versions involves additions, removals, and/or changes
to fields in your project's `PROJECT` file, which is created by running the `init` command.

The `PROJECT` file now has a new layout.  It stores more information about what resources are in use, to better enable plugins to make useful decisions when scaffolding.
    
Furthermore, the PROJECT file itself is now versioned: the `version` field corresponds to the version of the PROJECT file itself, while the `layout` field indicates the scaffolding & primary plugin version in use. 

### Steps to migrate

The following steps will describe the changes that needs to be applied to the project configuration file (`PROJECT`) that can be found at the root directory for you manually add the information that is used when Kubebuilder generate the files.  

- Add the `projectName`

The project name is the name of the project directory in lowercase:

```yaml
...
projectName: example
...
```

- Add the `layout`

The default plugin layout which is equivalent to the previous versions is `go.kubebuilder.io/v2`:

```yaml
...
layout: go.kubebuilder.io/v2
...
```

- Update the `version`

> The `version` field represents the version of Project layouts. So, you ought to update this to `"3"`:

Reviewer: Nope, that's not for Kubebuilder 3.x, that is related to the project configuration version, not to the library version.

> The `version` field represents the version of Project layouts. So, you ought to update this to `"3"`:

Reviewer 2: Can change it to something else, but a migration guide should be clear, not "you ought to". Either they need to change it or not. If there is a decision to be made, more information is needed about why each option might be picked.

> The `version` field represents the version of Project layouts. So, you ought to update this to `"3"`:

Author: The explanations are on top of the doc. However, I agree with your comment too. So, I changed it as well.

- Project version 2:

```yaml
domain: testproject.org
repo: sigs.k8s.io/kubebuilder/example
resources:
- group: crew
  kind: Captain
  version: v1
- group: crew
  kind: FirstMate
  version: v1
- group: crew
  kind: Admiral
  version: v1
version: "2"
```

- Project version 3:

```yaml
domain: testproject.org
layout: go.kubebuilder.io/v2
projectName: example
repo: sigs.k8s.io/kubebuilder/example
resources:
- api:
    crdVersion: v1
    namespaced: true
  controller: true
  domain: testproject.org
  group: crew
  kind: Captain
  path: example/api/v1
  version: v1
  webhooks:
    defaulting: true
    validation: true
    webhookVersion: v1
- api:
    crdVersion: v1
    namespaced: true
  controller: true
  domain: testproject.org
  group: crew
  kind: FirstMate
  path: example/api/v1
  version: v1
  webhooks:
    conversion: true
    webhookVersion: v1
- api:
    crdVersion: v1
  controller: true
  domain: testproject.org
  group: crew
  kind: Admiral
  path: example/api/v1
  plural: admirales
  version: v1
  webhooks:
    defaulting: true
    webhookVersion: v1
version: "3"
```

### Verification

> Finally, we can run `make` and `make docker-build` to ensure things are working

Author: No. Because we only applied the changes in the PROJECT file for the go/v2 which is related to the same previous layout.

kustomize: ## Download kustomize locally if necessary.
	$(call go-get-tool,$(KUSTOMIZE),sigs.k8s.io/kustomize/kustomize/v3@v3.8.7)

# go-get-tool will 'go get' any package $2 and install it to $1.
PROJECT_DIR := $(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
define go-get-tool
@[ -f $(1) ] || { \
set -e ;\
TMP_DIR=$$(mktemp -d) ;\
cd $$TMP_DIR ;\
go mod init tmp ;\
echo "Downloading $(2)" ;\
GOBIN=$(PROJECT_DIR)/bin go get $(2) ;\
rm -rf $$TMP_DIR ;\
}
endef
```

- To make your project use the `kustomize` version defined replace all `kustomize` target with `$(KUSTOMIZE)`

<aside class="note">
<h1>CRD and Webhooks API multi-version supportability</h1>

You can check all Makefile changes by looking in the samples projects generated in the `testdata` directory of the Kubebuilder repository.
</aside>

#### Update your controllers

<aside class="note warning">
<h1>Controller-runtime version updated has breaking changes</h1>

Check [sigs.k8s.io/controller-runtime release docs from 0.7.0+ version][controller-releases] for breaking changes.

</aside>

Replace:

```go 
func (r *<MyKind>Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
    ctx := context.Background() 

> 	log := r.Log.WithValues("cronjob", req.NamespacedName)

Reviewer: Actually, it should be the other way round. Tabs for all instead os spaces.

```

With:

```go 
func (r *<MyKind>Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {

> 	log := r.Log.WithValues("cronjob", req.NamespacedName)

Author: It is aligned now tks

```

#### Change Logger to use flag options

In the `main.go` file replace: 

```go
flag.Parse()

ctrl.SetLogger(zap.New(zap.UseDevMode(true)))
```

With:

```go
opts := zap.Options{

> 		Development: true,

Reviewer: We scaffold with tabs, we should be consistent, but replacing all spaces with tabs not the other way round.

> 		Development: true,

Reviewer 2: I defer to you on that, tab literals in documentation files get weird :) We should check how gitbook converts them in the end.

> 		Development: true,

Author: it is aligned

<h1>Path attribute</h1>

If you do not scaffold an API/resource and only generate a controller for the API(GKV) informed then, you do not need to add the path. 

Kubebuilder only supports work with core types and the APIs scaffolded in the project by unless you manually change the files you will be unable to work with external-types.  

However, the path will be always the import path used in your Go files to use the API.

</aside>

```yaml
...
resources:
- api:
    ...
  ...
  group: webapp
  kind: Guestbook
  path: example/api/v1
``` 

##### If your project is using webhooks then, add `resources[entry].webhooks.[type]: true` for each type generated and then, add `resources[entry].webhooks.webhookVersion: v1beta1`:

<aside class="note">
<h1>Webhooks attribute</h1>

The valid types are; `defaulting`, `validation` and `conversion`. You used the type to do its scaffold. 

And then, add the Kubernetes API version used to do the scaffolds in `Kubebuilder v2` was `v1beta1`.
 
</aside>

```yaml
resources:
- api:
    ...
  ...
  group: webapp
  kind: Guestbook
  webhooks:
    defaulting: true
    validation: true
    webhookVersion: v1beta1
```

#### Check your PROJECT file
 

> Now ensure that your `PROJECT` file us properly updated and have the full information which should be the same used when the files are create by the tool. See the `PROJECT file for the QuickStart example manually update to use `go.kubebuilder.io/v2` :

Reviewer 3: ```suggestion
Now ensure that your `PROJECT` file us properly updated and have the full information which should be the same used when the files are create by the tool. See the `PROJECT` file for the QuickStart example manually update to use `go.kubebuilder.io/v2` :
```

Reviewer 3: nit. Sorry for reporting it late :(

Reviewer 3: Otherwise, LGTM.

Author: Solved

# Migration from v2 to v3 by updating the files manually

Make sure you understand the [differences between Kubebuilder v2 and v3][migration-v2vsv3]
before continuing

Please ensure you have followed the [installation guide](/quick-start.md#installation)
to install the required components.

> The following guide will describe the manually steps required for
> you to upgrade only your config version and starts to use the plugins versions. 

Reviewer: Maybe?
```suggestion
The following guide describes the manual steps required to upgrade your config version and start using the plugin-enabled version.
```

This way is more complex, susceptible to errors, and success cannot be assured. Also, by following these steps you will not get the improvements and bug fixes in the default generated project files. 

Be aware that is recommended follow up the [Migration Guide v2 to V3][migration-guide-v2-to-v3] instead.

## Migration from project config version "2" to "3" 

Migrating between project configuration versions involves additions, removals, and/or changes
to fields in your project's `PROJECT` file, which is created by running the `init` command.

The `PROJECT` file now has a new layout.  It stores more information about what resources are in use, to better enable plugins to make useful decisions when scaffolding.
    
Furthermore, the `PROJECT` file itself is now versioned: the `version` field corresponds to the version of the `PROJECT` file itself, while the `layout` field indicates the scaffolding and the primary plugin version in use. 

### Steps to migrate

> The following steps will describe the changes that needs to be applied to the project configuration file (`PROJECT`) for you manually add the information that is used when Kubebuilder generate the files. Thi file can be found at the root directory. 

Reviewer: I think the first sentence is missing something.

...
```

#### Add the `layout`

The default plugin layout which is equivalent to the previous versions is `go.kubebuilder.io/v2`:

```yaml
...
layout: go.kubebuilder.io/v2
...
```

#### Update the `version`

The `version` field represents the version of Project layouts. Update this to `"3"`:

```yaml
...
version: "3"
...
```

#### Add the resource data
 
The attribute resources represent the list of resources scaffold in your project.

In this way, you will need to add the following data for each resource added to the project.

##### Add the Kubernetes API version by adding `resources[entry].api.crdVersion: v1beta1`: 

```yaml
...
resources:
- api:
    ...
    crdVersion: v1beta1
  domain: my.domain
  group: webapp
  kind: Guestbook
  ...
```
 
##### Add the scope used do scaffold the CRDs by adding `resources[entry].api.namespaced: true` unless they were cluster-scoped:

```yaml
...
resources:
- api:
    ...
    namespaced: true
  group: webapp
  kind: Guestbook
  ...
```

##### If you have a controller scaffolded for the API then, add `resources[entry].controller: true`:

```yaml
...
resources:
- api:
    ...
  controller: true  
  group: webapp
  kind: Guestbook
```

> ##### Add the resource domain such as `resources[entry].domain: testproject.org` which usually will be the project domain unless the API scaffold is a core type and/or an external type:

Reviewer: Generic external resources are not officially supported yet so if te user has one he will probably know what domain he needs to write.
Core resources may have `"k8s.io"` as the domain or the empty domain (field not present) and there is no apparent pattern to know which is the right one.

> ##### Add the resource domain such as `resources[entry].domain: testproject.org` which usually will be the project domain unless the API scaffold is a core type and/or an external type:

Author: I added further info with examples. Very detailed to documented it properly as has been requested for all points.

However, the path will be always the import path used in your Go files to use the API.

</aside>

```yaml
...
resources:
- api:
    ...
  ...
  group: webapp
  kind: Guestbook
  path: example/api/v1
``` 

##### If your project is using webhooks then, add `resources[entry].webhooks.[type]: true` for each type generated and then, add `resources[entry].webhooks.webhookVersion: v1beta1`:

<aside class="note">
<h1>Webhooks</h1>

The valid types are; `defaulting`, `validation` and `conversion`. You used the type to do its scaffold. 

The Kubernetes API version used to do the webhooks scaffolds in `Kubebuilder v2` is `v1beta1`. Then, you will add the `webhookVersion: v1beta1` for all cases. 

</aside>

```yaml
resources:
- api:
    ...
  ...
  group: webapp
  kind: Guestbook
  webhooks:
    defaulting: true
    validation: true
    webhookVersion: v1beta1
```

#### Check your PROJECT file
 
Now ensure that your `PROJECT` file has the full information which should be the same when the manifests are generated via Kubebuilder V3 CLI. 

> For the QuickStart example, the `PROJECT` file manually update to use `go.kubebuilder.io/v2` would like:

Reviewer: ```suggestion
For the QuickStart example, the `PROJECT` file manually updated to use `go.kubebuilder.io/v2` would look like:
```

	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
		"Enable leader election for controller manager. "+
			"Enabling this will ensure there is only one active controller manager.")
```

With: 

```go
func main() {
	var metricsAddr string
	var enableLeaderElection bool
	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
		"Enable leader election for controller manager. "+
			"Enabling this will ensure there is only one active controller manager.")
```

And then, rename the flags in the `config/default/manager_auth_proxy_patch.yaml` and `config/default/manager.yaml`:

```yaml
- name: manager
args:
- "--health-probe-bind-address=:8081"
- "--metrics-bind-address=127.0.0.1:8080"
- "--leader-elect"
```

#### Verification

Finally, we can run `make` and `make docker-build` to ensure things are working
fine.

## Change your project to remove the Kubernetes deprecated API versions usage

<aside class="note">
<h1>Before continuing</h1>

Make sure you understand [Versions in CustomResourceDefinitions][custom-resource-definition-versioning].

</aside>

> The following steps
> will describe a workflow option to upgrade your project in order to remove the Kubernetes API versions usage which are deprecated `apiextensions.k8s.io/v1beta1`, `admissionregistration.k8s.io/v1beta1`, `cert-manager.io/v1alpha2`.

Reviewer 4: ```suggestion
The following steps describe a workflow to upgrade your project to remove the deprecated Kubernetes APIs: `apiextensions.k8s.io/v1beta1`, `admissionregistration.k8s.io/v1beta1`, `cert-manager.io/v1alpha2`.
```

> Currently, the Kubebuilder CLI tool does not support scaffold resources for both Kubernetes API versions such as; an API/CRD with `apiextensions.k8s.io/v1beta1` and another one with `apiextensions.k8s.io/v1`.

Reviewer 4: ```suggestion
The Kubebuilder CLI tool does not support scaffolded resources for both Kubernetes API versions such as; an API/CRD with `apiextensions.k8s.io/v1beta1` and another one with `apiextensions.k8s.io/v1`.
```

<aside class="note">
<h1>Cert Manager API</h1>

If you scaffold a webhook using the Kubernetes API `admissionregistration.k8s.io/v1` then, by default, it will use the API `cert-manager.io/v1` in the manifests. 

</aside>

The first step is to update your `PROJECT` file by replacing the `api.crdVersion:v1beta` and `webhooks.WebhookVersion:v1beta` with `api.crdVersion:v1` and `webhooks.WebhookVersion:v1` which would look like: 

```yaml
domain: my.domain
layout: go.kubebuilder.io/v3
projectName: example
repo: example
resources:
- api:
    crdVersion: v1
    namespaced: true
  group: webapp
  kind: Guestbook
  version: v1
  webhooks:
    defaulting: true
    webhookVersion: v1
version: "3"
```

> Now, you can try to re-create the APIS(CRDs) and Webhooks manifests by running the commands `kubebuilder create api` and `kubebuilder create webhook` for the same group, kind and versions with the flag `--force`.   
> 
> However, note that the tool will re-scaffold the files which means that you will lose its content. So, before executing the commands ensure that you have its content stored in another place. An  easy option is to use `git` to compare your locally changes with the previous versions in order to recover its contents.  

Reviewer 4: The note warning of impeding file destruction should come before the commands that will do the destruction :) 

```suggestion
You can try to re-create the APIS(CRDs) and Webhooks manifests by using the `--force` flag.

Note, however, that the tool will re-scaffold the files which means that you will lose their content. Before executing the commands ensure that you have the files' content stored in another place. An easy option is to use `git` to compare your local change with the previous version to recover the contents.

Now, re-create the APIS(CRDs) and Webhooks manifests by running the  `kubebuilder create api` and `kubebuilder create webhook` for the same group, kind and versions with the flag `--force`, respectively.
```
